core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowBasePlugin.scala

```diff
@@ -0,0 +1,126 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cloudflow.sbt
+
+import java.nio.charset.StandardCharsets._
+import java.util.Base64
+import java.util.zip.Deflater
+import java.io.ByteArrayOutputStream
+
+import com.typesafe.config._
+
+import sbt._
+import sbt.Keys._
+import sbtdocker.DockerKeys._
+import com.typesafe.sbt.packager.archetypes._
+import spray.json._
+
+import cloudflow.sbt.CloudflowKeys._
+import cloudflow.blueprint.StreamletDescriptorFormat._
+import cloudflow.blueprint.StreamletDescriptor
+
+abstract class CloudflowBasePlugin extends AutoPlugin {
+  final val AppHome = "${app_home}"
+  final val AppTargetDir: String = "/app"
+  final val appTargetSubdir: String ⇒ String = dir ⇒ s"$AppTargetDir/$dir"
+  final val AppJarsDir: String = "app-jars"
+  final val DepJarsDir: String = "dep-jars"
+  final val OptAppDir = "/opt/cloudflow/"
+  final val scalaVersion = "2.12"
+
+  // NOTE !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+  // The UID and GID of the `jboss` user is used in different parts of Cloudflow
+  // If you change this, you have to make sure that all references to this value are changed
+  // - fsGroups on streamlet pods uses the GID to make volumes readable
+  val UserInImage = "185" // default non-root user in the spark image
+  val userAsOwner: String ⇒ String = usr ⇒ s"$usr:cloudflow"
+  val StreamletDescriptorsLabelName = "com.lightbend.cloudflow.streamlet-descriptors"
+
+  override def requires = CommonSettingsAndTasksPlugin && StreamletScannerPlugin &&
+    JavaAppPackaging && sbtdocker.DockerPlugin
+
+  // override def projectSettings = Seq ()
+
+  private[sbt] val verifyDockerRegistry = Def.task {
+    cloudflowDockerRegistry.value.getOrElse(throw DockerRegistryNotSet)
+  }
+
+  private[sbt] val checkUncommittedChanges = Def.task {
+    val log = streams.value.log
+    if (cloudflowBuildNumber.value.hasUncommittedChanges) {
+      log.warn(s"You have uncommitted changes in ${thisProjectRef.value.project}. Please commit all changes before publishing to guarantee a repeatable and traceable build.")
+    }
+  }
+
+  private[sbt] def makeStreamletDescriptorsLabelValue(streamletDescriptorsJson: JsValue) = {
+    // create a root object with the array
+    val streamletDescriptorsJsonStr =
+      JsObject("descriptors" -> streamletDescriptorsJson)
+        .compactPrint
+
+    val compressed = zlibCompression(streamletDescriptorsJsonStr.getBytes(UTF_8))
+    Base64.getEncoder.encodeToString(compressed)
+  }
+
+  private[sbt] val showResultOfBuild = Def.task {
+    val log = streams.value.log
+    val imagePushed = (imageNames in docker).value.head // assuming we only build a single image!
```

Comment:
Does this work correcly with sbt multi project (sub projects?)

Reply:
The plugin will be attached to one project (which becomes 1 sub-project in a multi-project build). So the image generation will be one sub-project at a time. If from the top level project we issue a `build` or `buildAndPublish` it will create one image per sub-project. This is what we need - right ?

Reply:
Yes, that sounds correct.